include/xtensor/xhistogram.hpp

```diff
@@ -0,0 +1,312 @@
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+/**
+ * @brief construct histogram
+ */
+
+#ifndef XTENSOR_HISTOGRAM_HPP
+#define XTENSOR_HISTOGRAM_HPP
+
+#include "xtensor.hpp"
+#include "xsort.hpp"
+
+namespace xt
+{
+    /**
+     * @ingroup histogram
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param data The data.
+     * @param bin_edges The bin-edges.
+     * @param weights Weight factors corresponding to each data-point.
+     * @param density If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2, class E3>
+    inline auto histogram(
+        const E1 &data, const E2 &bin_edges, const E3 &weights, bool density=false
+    )
+    {
+        // counter type
+        using size_type = detail::common_size_type_t<E1,E2,E3>;
+        using value_type = typename std::decay_t<E3>::value_type;
+
+        // basic checks
+        // - shape and size
+        XTENSOR_ASSERT( data     .dimension() == 1 );
+        XTENSOR_ASSERT( weights  .dimension() == 1 );
+        XTENSOR_ASSERT( bin_edges.dimension() == 1 );
+        XTENSOR_ASSERT( weights  .size() == data.size() );
+        XTENSOR_ASSERT( bin_edges.size() >= 2 );
+        // - bin-edges must be sorted
+        XTENSOR_ASSERT( std::is_sorted(bin_edges.cbegin(), bin_edges.cend()) );
+        // - data must be enclosed in the bins
+        XTENSOR_ASSERT( xt::amin(data)[0] >= bin_edges[0] );
+        XTENSOR_ASSERT( xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1] );
+
+        // initialize output
+        xt::xtensor<value_type,1> count = xt::zeros<value_type>({bin_edges.size()-1});
+
+        // indices that sort "data"
+        xt::xtensor<size_t,1> isort = xt::argsort(data);
+
+        // index of the current bin
+        size_type ibin = 0;
+
+        // fill the histogram: loop over (sorted) data
+        for (auto &idx : isort)
+        {
+            // - proceed to the relevant bin
+            while (data[idx] >= bin_edges[ibin+1] && ibin < bin_edges.size()-2)
+            {
+                ++ibin;
+            }
+            // - update the count
+            count[ibin] += weights[idx];
+        }
+
+        // cast type
+        xt::xtensor<double,1> P = xt::cast<double>(count);
+
+        // normalize
+        if ( density )
+        {
+            // - size as doubles
+            double N = static_cast<double>(data.size());
+            // - apply normalization
+            for (size_type i = 0 ; i < bin_edges.size()-1 ; ++i)
+            {
+                P[i] /= ( static_cast<double>(bin_edges[i+1] - bin_edges[i]) * N );
+            }
+        }
```

Comment:
Sorry for not being clear in my previous comment, I meant a double temporary:
```cpp
for(for (size_type i = 0 ; i < bin_edges.size()-1 ; ++i)
{
    double p = static_cast<double>(count[i]) / ((bin_edges[i+1] - bin_edges[i]) * N);
    count[i] = static_cast<value_type>(p);
}
```

Reply:
@JohanMabille I'm still not sure about any non-`double` return from `histogram`. I think that the most logical (but impossible in C++) behaviour would be:

* `density=false`: returns an integer array with the counts,
* `density=true`: returns an double array with the probabilities.

I don't think the type of `data` or `weights` should weigh-in here. In that case I believe that the current behaviour that always returns a double array is most robust.

Reply:
Indeed you cannot have a return type depending on a dynamic value. So the idea was to make the return type similar to the one passed as an argument, expecting `weights` to be an array of integers when `density=false` (but I might be wrong on that point).

The idea is to avoid to impose the `double` type to the user (she may want to use float or hand-made higher precision floating point type).

Another solution is to let the user specify its return type (and provide `double` as default value):
```cpp
template<cmass R = double, class E1, class E2, class E3>
inline auto histogram(const E1 &data, const E2 &bin_edges, const E3 &weights, bool density=false
```

Reply:
@JohanMabille I now see your point. I'm fine with both, though I have a slight preference from the return template. If you have a strong preference I will will follow yours.

Reply:
Let's go for the return template then. It is likely to be consistent with the future fix of #1106 for instance.